// react-table/lib/hooks/useExpanded.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useExpanded = undefined;

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }();

var _react = require("react");

var _utils = require("../utils");

var _actions = require("../actions");

var _useTableState = require("./useTableState");

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

_useTableState.defaultState.expanded = {};

(0, _actions.addActions)({
  toggleExpanded: "__toggleExpanded__",
  useExpanded: "__useExpanded__"
});

var useExpanded = exports.useExpanded = function useExpanded(props) {
  var debug = props.debug,
      columns = props.columns,
      rows = props.rows,
      _props$expandedKey = props.expandedKey,
      expandedKey = _props$expandedKey === undefined ? "expanded" : _props$expandedKey,
      hooks = props.hooks,
      _props$state = _slicedToArray(props.state, 2),
      expanded = _props$state[0].expanded,
      setState = _props$state[1];

  var toggleExpandedByPath = function toggleExpandedByPath(path, set) {
    return setState(function (old) {
      var expanded = old.expanded;

      var existing = (0, _utils.getBy)(expanded, path);
      set = (0, _utils.getFirstDefined)(set, !existing);
      return _extends({}, old, {
        expanded: (0, _utils.setBy)(expanded, path, set)
      });
    }, _actions.actions.toggleExpanded);
  };

  hooks.row.push(function (row) {
    var path = row.path;

    row.toggleExpanded = function (set) {
      return toggleExpandedByPath(path, set);
    };
  });

  var expandedRows = (0, _react.useMemo)(function () {
    if (debug) console.info("getExpandedRows");

    var expandedRows = [];

    // Here we do some mutation, but it's the last stage in the
    // immutable process so this is safe
    var handleRow = function handleRow(row, index) {
      var depth = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 0;
      var parentPath = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : [];

      // Compute some final state for the row
      var path = [].concat(_toConsumableArray(parentPath), [index]);

      row.path = path;
      row.depth = depth;

      row.isExpanded = row.original && row.original[expandedKey] || (0, _utils.getBy)(expanded, path);

      row.cells = columns.map(function (column) {
        var cell = {
          column: column,
          row: row,
          state: null,
          value: row.values[column.id]
        };

        return cell;
      });

      expandedRows.push(row);

      if (row.isExpanded && row.subRows && row.subRows.length) {
        row.subRows.forEach(function (row, i) {
          return handleRow(row, i, depth + 1, path);
        });
      }
    };

    rows.forEach(function (row, i) {
      return handleRow(row, i);
    });

    return expandedRows;
  }, [rows, expanded, columns]);

  var expandedDepth = findExpandedDepth(expanded);

  return _extends({}, props, {
    toggleExpandedByPath: toggleExpandedByPath,
    expandedDepth: expandedDepth,
    rows: expandedRows
  });
};

function findExpandedDepth(obj) {
  var depth = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 1;

  return Object.values(obj).reduce(function (prev, curr) {
    if ((typeof curr === "undefined" ? "undefined" : _typeof(curr)) === "object") {
      return Math.max(prev, findExpandedDepth(curr, depth + 1));
    }
    return depth;
  }, 0);
}
