// react-table/lib/hooks/useRows.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useRows = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require("react");

var useRows = exports.useRows = function useRows(props) {
  var debug = props.debug,
      columns = props.columns,
      _props$subRowsKey = props.subRowsKey,
      subRowsKey = _props$subRowsKey === undefined ? "subRows" : _props$subRowsKey,
      data = props.data;


  var accessedRows = (0, _react.useMemo)(function () {
    if (debug) console.info("getAccessedRows");

    // Access the row's data
    var accessRow = function accessRow(originalRow, i) {
      var depth = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 0;

      // Keep the original reference around
      var original = originalRow;

      // Process any subRows
      var subRows = originalRow[subRowsKey] ? originalRow[subRowsKey].map(function (d, i) {
        return accessRow(d, i, depth + 1);
      }) : undefined;

      var row = {
        original: original,
        index: i,
        subRows: subRows,
        depth: depth
      };

      // Create the cells and values
      row.values = {};
      columns.forEach(function (column) {
        row.values[column.id] = column.accessor ? column.accessor(originalRow, i, { subRows: subRows, depth: depth, data: data }) : undefined;
      });

      return row;
    };

    // Use the resolved data
    return data.map(function (d, i) {
      return accessRow(d, i);
    });
  }, [data, columns]);

  return _extends({}, props, {
    rows: accessedRows
  });
};
