// react-table/es/hooks/useTable.js
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

import PropTypes from 'prop-types';
//
import { flexRender, applyHooks, applyPropHooks, mergeProps } from '../utils';

import { useTableState } from './useTableState';

var renderErr = 'You must specify a render "type". This could be "Header", "Filter", or any other custom renderers you have set on your column.';

var propTypes = {
  // General
  data: PropTypes.any,
  columns: PropTypes.arrayOf(PropTypes.shape({
    aggregate: PropTypes.func,
    filterFn: PropTypes.func,
    filterAll: PropTypes.bool,
    sortByFn: PropTypes.func,
    resolvedDefaultSortDesc: PropTypes.bool,
    canSortBy: PropTypes.bool,
    canGroupBy: PropTypes.bool,
    Cell: PropTypes.any,
    Header: PropTypes.any,
    Filter: PropTypes.any
  })),

  filterFn: PropTypes.func,
  sortByFn: PropTypes.func,
  orderByFn: PropTypes.func,
  groupByFn: PropTypes.func,

  manualGrouping: PropTypes.bool,
  manualFilters: PropTypes.bool,
  manualSorting: PropTypes.bool,

  defaultSortDesc: PropTypes.bool,
  disableMultiSort: PropTypes.bool,
  subRowsKey: PropTypes.string,
  expandedKey: PropTypes.string,
  userAggregations: PropTypes.object,

  debug: PropTypes.bool
};

export var useTable = function useTable(props) {
  for (var _len = arguments.length, plugins = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
    plugins[_key - 1] = arguments[_key];
  }

  // Validate props
  PropTypes.checkPropTypes(propTypes, props, 'property', 'useTable');

  // Destructure props
  var _props$data = props.data,
      data = _props$data === undefined ? [] : _props$data,
      userState = props.state,
      debug = props.debug;

  // Always provide a default state

  var defaultState = useTableState();

  // But use the users state if provided
  var state = userState || defaultState;

  // These are hooks that plugins can use right before render
  var hooks = {
    beforeRender: [],
    columns: [],
    headers: [],
    headerGroups: [],
    rows: [],
    row: [],
    renderableRows: [],
    getTableProps: [],
    getRowProps: [],
    getHeaderRowProps: [],
    getHeaderProps: [],
    getCellProps: []

    // The initial api
  };var api = _extends({}, props, {
    data: data,
    state: state,
    hooks: hooks
  });

  if (debug) console.time('hooks');
  // Loop through plugins to build the api out
  api = plugins.filter(Boolean).reduce(function (prev, next) {
    return next(prev);
  }, api);
  if (debug) console.timeEnd('hooks');

  // Run the beforeRender hook
  if (debug) console.time('hooks.beforeRender');
  applyHooks(api.hooks.beforeRender, undefined, api);
  if (debug) console.timeEnd('hooks.beforeRender');

  if (debug) console.time('hooks.columns');
  api.columns = applyHooks(api.hooks.columns, api.columns, api);
  if (debug) console.timeEnd('hooks.columns');

  if (debug) console.time('hooks.headers');
  api.headers = applyHooks(api.hooks.headers, api.headers, api);
  if (debug) console.timeEnd('hooks.headers');
  [].concat(_toConsumableArray(api.columns), _toConsumableArray(api.headers)).forEach(function (column) {
    // Give columns/headers rendering power
    column.render = function (type) {
      var userProps = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

      if (!type) {
        throw new Error(renderErr);
      }
      return flexRender(column[type], _extends({}, api, column, userProps));
    };

    // Give columns/headers getHeaderProps
    column.getHeaderProps = function (props) {
      return mergeProps({
        key: ['header', column.id].join('_')
      }, applyPropHooks(api.hooks.getHeaderProps, column, api), props);
    };
  });

  if (debug) console.time('hooks.headerGroups');
  api.headerGroups = applyHooks(api.hooks.headerGroups, api.headerGroups, api).filter(function (headerGroup, i) {
    // Filter out any headers and headerGroups that don't have visible columns
    headerGroup.headers = headerGroup.headers.filter(function (header) {
      var recurse = function recurse(columns) {
        return columns.filter(function (column) {
          if (column.columns) {
            return recurse(column.columns);
          }
          return column.visible;
        }).length;
      };
      if (header.columns) {
        return recurse(header.columns);
      }
      return header.visible;
    });

    // Give headerGroups getRowProps
    if (headerGroup.headers.length) {
      headerGroup.getRowProps = function () {
        var props = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
        return mergeProps({
          key: ['header' + i].join('_')
        }, applyPropHooks(api.hooks.getHeaderRowProps, headerGroup, api), props);
      };
      return true;
    }

    return false;
  });
  if (debug) console.timeEnd('hooks.headerGroups');

  // Run the rows (this could be a dangerous hook with a ton of data)
  if (debug) console.time('hooks.rows');
  api.rows = applyHooks(api.hooks.rows, api.rows, api);
  if (debug) console.timeEnd('hooks.rows');

  // This function is absolutely necessary and MUST be called on
  // any rows the user wishes to be displayed.
  api.prepareRow = function (row) {
    var index = row.index;

    row.getRowProps = function (props) {
      return mergeProps({ key: ['row', index].join('_') }, applyHooks(api.hooks.getRowProps, row, api), props);
    };

    row.cells = row.cells.filter(function (cell) {
      return cell.column.visible;
    });

    row.cells.forEach(function (cell) {
      if (!cell) {
        return;
      }

      var column = cell.column;


      cell.getCellProps = function (props) {
        var columnPathStr = [index, column.id].join('_');
        return mergeProps({
          key: ['cell', columnPathStr].join('_')
        }, applyPropHooks(api.hooks.getCellProps, cell, api), props);
      };

      cell.render = function (type) {
        var userProps = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

        if (!type) {
          throw new Error('You must specify a render "type". This could be "Cell", "Header", "Filter", "Aggregated" or any other custom renderers you have set on your column.');
        }
        return flexRender(column[type], _extends({}, api, cell, userProps));
      };
    });
  };

  api.getTableProps = function (userProps) {
    return mergeProps(applyPropHooks(api.hooks.getTableProps, api), userProps);
  };

  api.getRowProps = function (userProps) {
    return mergeProps(applyPropHooks(api.hooks.getRowProps, api), userProps);
  };

  return api;
};
